pkg/plugins/runtime/k8s/controllers/gatewayapi/http_route_controller.go

```diff
@@ -83,32 +84,65 @@ func (r *HTTPRouteReconciler) gapiToKumaRoutes(
 	ParentConditions,
 	error,
 ) {
+	routeNs := kube_core.Namespace{}
+	if err := r.Client.Get(ctx, kube_types.NamespacedName{Name: route.Namespace}, &routeNs); err != nil {
+		if kube_apierrs.IsNotFound(err) {
+			return nil, nil, nil
+		} else {
+			return nil, nil, err
+		}
+	}
+
 	var refRoutes []gatewayapi.ParentRef
+	conditions := ParentConditions{}
 
 	kumaRoute := &mesh_proto.GatewayRoute{}
 
 	// Convert GAPI parent refs into selectors
 	for i, ref := range route.Spec.ParentRefs {
-		if handle, err := r.shouldHandleParentRef(ctx, route, ref); err != nil {
+		refAttachment, err := attachment.ParentRefAttachment(ctx, r.Client, &routeNs, ref)
+		if err != nil {
 			return nil, nil, errors.Wrapf(err, "unable to check parent ref %d", i)
-		} else if !handle {
-			continue
 		}
-		refRoutes = append(refRoutes, ref)
 
-		refNamespace := route.Namespace
-		if ns := ref.Namespace; ns != nil {
-			refNamespace = string(*ns)
+		switch refAttachment {
+		// TODO are these three cases reflected correctly?
+		// specifically does ResolvedRefs cover ParentRefs or only BackendRefs?
+		case attachment.NotPermitted, attachment.Invalid:
+			var message string
+			switch refAttachment {
+			case attachment.NotPermitted:
+				message = "attachment to parent not permitted by AllowedRoutes"
+			case attachment.Invalid:
+				// TODO missing a specific Reason for this?
+				message = "reference to parent is invalid"
+			}
+			conditions[ref] = []kube_meta.Condition{
+				{
+					Type:    string(gatewayapi.ConditionRouteResolvedRefs),
+					Status:  kube_meta.ConditionFalse,
+					Reason:  string(gatewayapi.ListenerReasonRefNotPermitted),
+					Message: message,
+				},
+				{
+					Type:    string(gatewayapi.ConditionRouteAccepted),
+					Status:  kube_meta.ConditionFalse,
+					Reason:  "Refused", // kubernetes-sigs/gateway-api#972
+					Message: string(gatewayapi.ListenerReasonRefNotPermitted),
+				},
```

Comment:
Agreed, I've been avoiding adding helpers around the conditions but it's very messy. I would also want to introduce a subset of `Condition`'s fields to avoid the ambiguity around responsibility for setting `ObservedGeneration`, `LastTransition`, etc.

Reply:
https://github.com/kumahq/kuma/issues/3714